### PR TITLE
Kaqqao parsed directive arguments

### PR DESCRIPTION
**Pull request by bbakerman**
This is related to #1229 - I could not work out how to push to @kaqqao original branch - so I made my own

The original issue is #1228

This takes his ideas and uses GraphqlDirective as the runtime object for directives and arguments rather than `Map<String,Map<String,Object>` representation.

So now given

```
{
    books(author: "John Doe") @cache(enabled = true) {
        isbn
        title
    }
}
```

a data fetcher can do this

```
DataFetcher df =  env -> {
   GraphqlArgument arg = env.getDirectiveArgument("cache","enabled)
   boolean enabled = arg == null ? false : Boolean.valueOf(arg.getValue())
     
}
```

### CODE REVIEW

File: src/main/java/graphql/analysis/MaxQueryComplexityInstrumentation.java [hunk 1]
```diff
@@ -21,7 +21,7 @@
 /**
  * Prevents execution if the query complexity is greater than the specified maxComplexity.
  *
- * Use the {@link Function<QueryComplexityInfo, Boolean>} parameter to supply a function to perform a custom action when the max complexity
+ * Use the `Function&lt;QueryComplexityInfo Boolean&gt;` parameter to supply a function to perform a custom action when the max complexity
  * is exceeded. If the function returns {@code true} a {@link AbortExecutionException} is thrown.
  */
```
Comment (bbakerman) [inline src/main/java/graphql/analysis/MaxQueryComplexityInstrumentation.java]: This was stopping the command line version from building ebacuse of JavaDoc errors.  Clearly our builds dont run the same profile of JavaDoc as I do.  Not sure why

File: src/main/java/graphql/execution/ConditionalNodes.java [hunk 2]
```diff
@@ -10,6 +11,7 @@
 import static graphql.language.NodeUtil.directivesByName;
```
Comment (bbakerman) [inline src/main/java/graphql/execution/ConditionalNodes.java]: Broken windows fix

File: src/main/java/graphql/execution/DirectivesResolver.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    private final ValuesResolver valuesResolver;
+
+    public DirectivesResolver(ValuesResolver valuesResolver) {
+        this.valuesResolver = valuesResolver;
+    }
+
+    public Map<String, GraphQLDirective> getFieldDirectives(Field field, GraphQLSchema schema, Map<String, Object> variables) {
+        GraphqlFieldVisibility fieldVisibility = schema.getFieldVisibility();
+
+        Map<String, GraphQLDirective> directiveMap = new HashMap<>();
```
Comment (andimarek) [inline src/main/java/graphql/execution/DirectivesResolver.java]: LinkedHasMap

File: src/main/java/graphql/schema/DataFetchingEnvironmentImpl.java [hunk 4]
```diff
@@ -84,6 +88,22 @@ public boolean containsArgument(String name) {
         return (T) arguments.get(name);
     }
 
+
+    @Override
+    public Map<String, GraphQLDirective> getDirectives() {
+        return new HashMap<>(directives);
```
Comment (andimarek) [inline src/main/java/graphql/schema/DataFetchingEnvironmentImpl.java]: LinkedHashMap

**Comment by bbakerman** [comment]
@andimarek - there is an integration test inside `graphql.execution.DirectivesResolverTest`

**Comment by bbakerman** [comment]
> Gave this a bit more thought, and I think there should be a method that returns a map of parsed directives for any directive container. This is because one might need to get directives from a different field than the first (if DataFetchingEnvironment#getFields contains multiple fields) or from a fragment or fragment definition.

> I can do it if you agree on the approach.

@kaqqao - what shape do you see for this?  as in what is the key?  The Field node?  Should be not just combine them??

Do you need the AST nodes per set of directives?

```
{
    books(author: "John Doe") @cache(enabled = true) {
        isbn
        title
    }
   also : ...FragX
}
 
fragment FragX {
   books @burnAfterReading {
     title
   }
}
```

**Comment by kaqqao** [comment]
The way I have it right now is that I return a new object (called `Directives`) that is a glorified `Map<Introspection.DirectiveLocation, Map<String, List<GraphQLDirective>>`. 

I.e. it is a map of directives by name by location. The reason it's a `List` is because the same directive (same name) could appear in multiple places but with potentially different argument values. So the directives for both `books` fields in your example are merged inside the inner `Map<String, List<GraphQLDirective>>`, they're not keyed by the AST node. Not sure if that's the way it _should_ be, that's just the way I have it now...

I then have some convenience methods on the `Directives` class for getting a single directive or the whole list per name per location.

**Comment by kaqqao** [comment]
I added a comprehensive example and more rationale to the previous message, so if you're reading this from an email, please make sure to see the updated version of it.

**Comment by bbakerman** [comment]
This has now been reverted via https://github.com/graphql-java/graphql-java/pull/1329

@kaqqao - if you want to have a go at implenting it again (starting with the original code as  a base) then please do so.

**Pull request merged**